Clang: Crash on capturing invalid structured binding declaration in lambda expression (regression from 44f2baa3804a62)

The following code ICE's (crash on invalid) on trunk, and going back to at least clang 16.0.6

```c++
struct Point {
   int x;
   int y;
};

Point get_point() { return { 1, 2 }; }

int main() {
    auto& [x, y] = get_point();

    bool b = [x, y] {
        return true;
    }();

    return b;
}
```

Godbolt link: https://godbolt.org/z/bvEbbc6hj

The compiler correctly diagnoses that we need to use `auto&&` for these variables if we want to use them in a structured binding declaration of an rvalue.... and then crashes.

 I ran into this refactoring some code that manually extracted these members (that were already captured in a lambda expression) to use a structured binding. This made clangd crash outright instead of letting me know that I had an error.

> In `SemaDecl.cpp` L13052, both GDB and LLDB think the variable is an InvalidDecl.

I don't get that behavior in MSVC's debugger. `VDecl->isInvalidDecl()` returns `false` for me. For me, it gets diagnosed when performing the initialization: https://github.com/llvm/llvm-project/blob/918863deb0af1986de2df96ca6094eb625adafed/clang/lib/Sema/SemaInit.cpp#L5089

so it issues:
```
C:\Users\aballman\OneDrive - Intel Corporation\Desktop\test.cpp:9:11: error: non-const lvalue reference to type 'Point' cannot bind to a temporary of type 'Point'
    9 |     auto& [x, y] = get_point();
      |           ^        ~~~~~~~~~~~
```
then it creates a recovery expression and pretends the variable was initialized from that: https://github.com/llvm/llvm-project/blob/f3c417f341135ff357f2d454866d7d05262ebfca/clang/lib/Sema/SemaDecl.cpp#L13515 (note, `VDecl->InvalidDecl` is still `false` at this point for me).

When I call `setInvalidDecl()` from there, the later crash is gone, but I also get a bunch of test failures from the test suite:
```
F:\source\llvm-project>python llvm\out\build\x64-Debug\bin\llvm-lit.py -sv -j61 clang\test\
llvm-lit.py: F:\source\llvm-project\llvm\utils\lit\lit\llvm\config.py:46: note: using lit tools: C:\GnuWin32\bin
llvm-lit.py: F:\source\llvm-project\llvm\utils\lit\lit\llvm\config.py:487: note: using clang: f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe
llvm-lit.py: F:\source\llvm-project\llvm\utils\lit\lit\discovery.py:250: warning: test suite 'Clang-Unit' contained no tests
-- Testing: 17988 tests, 61 workers --
Testing:  0.. 10.. 20.. 30.. 40.. 50..
FAIL: Clang :: OpenMP/task_messages.cpp (1258 of 17988)
******************** TEST 'Clang :: OpenMP/task_messages.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -verify=expected,omp45 -fopenmp-version=45 -fopenmp -ferror-limit 200 -std=c++11 -o - F:\source\llvm-project\clang\test\OpenMP\task_messages.cpp -Wuninitialized
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -verify=expected,omp45 -fopenmp-version=45 -fopenmp -ferror-limit 200 -std=c++11 -o - 'F:\source\llvm-project\clang\test\OpenMP\task_messages.cpp' -Wuninitialized
# .---command stderr------------
# | error: 'expected-note' diagnostics seen but not expected:
# |   File F:\source\llvm-project\clang\test\OpenMP\task_messages.cpp Line 58: predetermined as a firstprivate in a task construct here
# |   File F:\source\llvm-project\clang\test\OpenMP\task_messages.cpp Line 58: predetermined as a firstprivate in a task construct here
# | 2 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60
FAIL: Clang :: SemaCXX/cxx2a-explicit-bool.cpp (1463 of 17988)
******************** TEST 'Clang :: SemaCXX/cxx2a-explicit-bool.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -std=c++17 -fsyntax-only F:\source\llvm-project\clang\test\SemaCXX\cxx2a-explicit-bool.cpp -verify=expected,pre20 -Wno-c++2a-extensions
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -std=c++17 -fsyntax-only 'F:\source\llvm-project\clang\test\SemaCXX\cxx2a-explicit-bool.cpp' -verify=expected,pre20 -Wno-c++2a-extensions
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaCXX\cxx2a-explicit-bool.cpp Line 128: no viable overloaded '='
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.
FAIL: Clang :: C/drs/dr0xx.c (2126 of 17988)
******************** TEST 'Clang :: C/drs/dr0xx.c' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -std=c89 -fsyntax-only -verify=expected,c89only -pedantic -Wno-declaration-after-statement -Wno-c11-extensions F:\source\llvm-project\clang\test\C\drs\dr0xx.c
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -std=c89 -fsyntax-only -verify=expected,c89only -pedantic -Wno-declaration-after-statement -Wno-c11-extensions 'F:\source\llvm-project\clang\test\C\drs\dr0xx.c'
# .---command stderr------------
# | error: 'c89only-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\C\drs\dr0xx.c Line 487: assigning to 'struct Two' from incompatible type 'struct One'
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70
FAIL: Clang :: SemaCXX/constexpr-default-init-value-crash.cpp (4384 of 17988)
******************** TEST 'Clang :: SemaCXX/constexpr-default-init-value-crash.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc F:\source\llvm-project\clang\test\SemaCXX\constexpr-default-init-value-crash.cpp -std=c++20 -fsyntax-only -verify
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc 'F:\source\llvm-project\clang\test\SemaCXX\constexpr-default-init-value-crash.cpp' -std=c++20 -fsyntax-only -verify
# RUN: at line 2
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc F:\source\llvm-project\clang\test\SemaCXX\constexpr-default-init-value-crash.cpp -std=c++20 -fsyntax-only -verify -fno-recovery-ast
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc 'F:\source\llvm-project\clang\test\SemaCXX\constexpr-default-init-value-crash.cpp' -std=c++20 -fsyntax-only -verify -fno-recovery-ast
# .---command stderr------------
# | error: 'expected-error' diagnostics seen but not expected:
# |   File F:\source\llvm-project\clang\test\SemaCXX\constexpr-default-init-value-crash.cpp Line 13: no return statement in constexpr function
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70..
FAIL: Clang :: SemaObjCXX/parameterized_classes_arc.mm (8129 of 17988)
******************** TEST 'Clang :: SemaObjCXX/parameterized_classes_arc.mm' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -fblocks -fobjc-arc -fobjc-runtime-has-weak F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm -verify
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -fblocks -fobjc-arc -fobjc-runtime-has-weak 'F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm' -verify
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 35: cannot initialize a parameter of type 'Forward *' with an lvalue of type 'Forward2 *__strong'
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 40: cannot initialize a parameter of type 'Forward *'
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 45: cannot initialize a parameter of type 'Forward *'
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 58: cannot initialize a parameter of type 'Forward *'
# | error: 'expected-note' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 14: passing argument to
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 14: passing argument to
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 14: passing argument to
# |   File F:\source\llvm-project\clang\test\SemaObjCXX\parameterized_classes_arc.mm Line 14: passing argument to
# | 8 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70..
FAIL: Clang :: SemaCXX/member-init.cpp (8182 of 17988)
******************** TEST 'Clang :: SemaCXX/member-init.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -fsyntax-only -fcxx-exceptions -verify -std=c++11 -Wall F:\source\llvm-project\clang\test\SemaCXX\member-init.cpp
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -fsyntax-only -fcxx-exceptions -verify -std=c++11 -Wall 'F:\source\llvm-project\clang\test\SemaCXX\member-init.cpp'
# .---command stderr------------
# | error: 'expected-warning' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaCXX\member-init.cpp Line 67: unused variable
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70..
FAIL: Clang :: C/C2x/n3042.c (9205 of 17988)
******************** TEST 'Clang :: C/C2x/n3042.c' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -verify -ffreestanding -Wno-unused -std=c2x F:\source\llvm-project\clang\test\C\C2x\n3042.c
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -verify -ffreestanding -Wno-unused -std=c2x 'F:\source\llvm-project\clang\test\C\C2x\n3042.c'
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\C\C2x\n3042.c Line 155: assigning to 'int' from incompatible type 'nullptr_t'
# |   File F:\source\llvm-project\clang\test\C\C2x\n3042.c Line 156: assigning to 'float' from incompatible type 'nullptr_t'
# |   File F:\source\llvm-project\clang\test\C\C2x\n3042.c Line 157: assigning to 'nullptr_t' from incompatible type 'int'
# |   File F:\source\llvm-project\clang\test\C\C2x\n3042.c Line 158: assigning to 'nullptr_t' from incompatible type 'float'
# | 4 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80
FAIL: Clang :: SemaCXX/block-call.cpp (10806 of 17988)
******************** TEST 'Clang :: SemaCXX/block-call.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -x c++ -fsyntax-only -verify F:\source\llvm-project\clang\test\SemaCXX\block-call.cpp -fblocks
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -x c++ -fsyntax-only -verify 'F:\source\llvm-project\clang\test\SemaCXX\block-call.cpp' -fblocks
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaCXX\block-call.cpp Line 34: invalid block pointer conversion assigning to 'int *(^)()' from 'int'
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80
FAIL: Clang :: C/C11/n1316.c (13137 of 17988)
******************** TEST 'Clang :: C/C11/n1316.c' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -verify F:\source\llvm-project\clang\test\C\C11\n1316.c
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -verify 'F:\source\llvm-project\clang\test\C\C11\n1316.c'
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 14: passing 'float' to parameter of incompatible type 'void *'
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 15: passing 'void *' to parameter of incompatible type 'float'
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 17: assigning to 'void *' from incompatible type 'float'
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 18: assigning to 'float' from incompatible type 'void *'
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 23: initializing 'void *' with an expression of incompatible type 'float'
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 23: initializing 'float' with an expression of incompatible type 'void *'
# | error: 'expected-note' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 7: passing argument to parameter here
# |   File F:\source\llvm-project\clang\test\C\C11\n1316.c Line 8: passing argument to parameter here
# | 8 errors generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80
FAIL: Clang :: SemaCXX/auto-invalid-init-crash.cpp (13729 of 17988)
******************** TEST 'Clang :: SemaCXX/auto-invalid-init-crash.cpp' FAILED ********************
Exit Code: 1

Command Output (stdout):
--
# RUN: at line 1
f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe -cc1 -internal-isystem f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include -nostdsysteminc -std=c++11 -fsyntax-only -fno-recovery-ast -verify F:\source\llvm-project\clang\test\SemaCXX\auto-invalid-init-crash.cpp
# executed command: 'f:\source\llvm-project\llvm\out\build\x64-debug\bin\clang.exe' -cc1 -internal-isystem 'f:\source\llvm-project\llvm\out\build\x64-debug\lib\clang\18\include' -nostdsysteminc -std=c++11 -fsyntax-only -fno-recovery-ast -verify 'F:\source\llvm-project\clang\test\SemaCXX\auto-invalid-init-crash.cpp'
# .---command stderr------------
# | error: 'expected-error' diagnostics expected but not seen:
# |   File F:\source\llvm-project\clang\test\SemaCXX\auto-invalid-init-crash.cpp Line 7: 'auto' deduced as 'int'
# | 1 error generated.
# `-----------------------------
# error: command failed with exit status: 1

--

********************
Testing:  0.. 10.. 20.. 30.. 40.. 50.. 60.. 70.. 80.. 90..
********************
Failed Tests (10):
  Clang :: C/C11/n1316.c
  Clang :: C/C2x/n3042.c
  Clang :: C/drs/dr0xx.c
  Clang :: OpenMP/task_messages.cpp
  Clang :: SemaCXX/auto-invalid-init-crash.cpp
  Clang :: SemaCXX/block-call.cpp
  Clang :: SemaCXX/constexpr-default-init-value-crash.cpp
  Clang :: SemaCXX/cxx2a-explicit-bool.cpp
  Clang :: SemaCXX/member-init.cpp
  Clang :: SemaObjCXX/parameterized_classes_arc.mm
```

--- a/clang/docs/ReleaseNotes.rst
+++ b/clang/docs/ReleaseNotes.rst
@@ -765,6 +765,9 @@ Bug Fixes in This Version
   Fixes (`#77583 <https://github.com/llvm/llvm-project/issues/77583>`_)
 - Fix an issue where CTAD fails for function-type/array-type arguments.
   Fixes (`#51710 <https://github.com/llvm/llvm-project/issues/51710>`_)
+- Fix crashes when using the binding decl from an invalid structured binding.
+  Fixes (`#67495 <https://github.com/llvm/llvm-project/issues/67495>`_) and
+  (`#72198 <https://github.com/llvm/llvm-project/issues/72198>`_)
 
 Bug Fixes to Compiler Builtins
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -13609,7 +13609,7 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit) {
       //  2) if the var decl is decompsition decl with a non-deduced type, and
       //     the initialization fails (e.g. `int [a] = {1, 2};`);
       // Case 1) was already handled elsewhere.
-      if (llvm::isa<DecompositionDecl>(VDecl)) // Case 2)
+      if (isa<DecompositionDecl>(VDecl)) // Case 2)
         VDecl->setInvalidDecl();
       return;
     }

--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -13606,9 +13606,9 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit) {
       // part in the valid bit of the declaration. There are a few exceptions:
       //  1) if the var decl has a deduced auto type, and the type cannot be
       //     deduced by an invalid initializer;
-      //  2) if the var decl is decompsition decl with a concrete type (e.g.
-      //    `int [a, b] = 1;`), and the initializer is invalid;
-      // Case 1) is already handled earlier in this function.
+      //  2) if the var decl is decompsition decl with a non-deduced type, and
+      //     the initialization fails (e.g. `int [a] = {1, 2};`);
+      // Case 1) was already handled elsewhere.
       if (llvm::isa<DecompositionDecl>(VDecl)) // Case 2)
         VDecl->setInvalidDecl();
       return;

--- a/clang/lib/Sema/SemaDecl.cpp
+++ b/clang/lib/Sema/SemaDecl.cpp
@@ -13602,6 +13602,15 @@ void Sema::AddInitializerToDecl(Decl *RealDecl, Expr *Init, bool DirectInit) {
           CreateRecoveryExpr(Init->getBeginLoc(), Init->getEndLoc(), Args);
       if (RecoveryExpr.get())
         VDecl->setInit(RecoveryExpr.get());
+      // In general, for error recovery purposes, the initalizer doesn't play
+      // part in the valid bit of the declaration. There are a few exceptions:
+      //  1) if the var decl has a deduced auto type, and the type cannot be
+      //     deduced by an invalid initializer;
+      //  2) if the var decl is decompsition decl with a concrete type (e.g.
+      //    `int [a, b] = 1;`), and the initializer is invalid;
+      // Case 1) is already handled earlier in this function.
+      if (llvm::isa<DecompositionDecl>(VDecl)) // Case 2)
+        VDecl->setInvalidDecl();
       return;
     }
 

--- a/clang/test/AST/ast-dump-invalid-initialized.cpp
+++ b/clang/test/AST/ast-dump-invalid-initialized.cpp
@@ -25,3 +25,16 @@ void test() {
   // CHECK: `-VarDecl {{.*}} invalid b5 'auto'
   auto b5 = A{1};
 }
+
+void pr72198() {
+  // CHECK: DecompositionDecl {{.*}} invalid 'int'
+  int [_, b] = {0, 0};
+  [b]{};
+}
+
+int get_point();
+void pr67495() {
+  // CHECK: DecompositionDecl {{.*}} invalid 'int &'
+  auto& [x, y] = get_point();
+  [x, y] {};
+}

--- a/clang/test/AST/ast-dump-invalid-initialized.cpp
+++ b/clang/test/AST/ast-dump-invalid-initialized.cpp
@@ -26,15 +26,17 @@ void test() {
   auto b5 = A{1};
 }
 
-void pr72198() {
+void GH72198() {
   // CHECK: DecompositionDecl {{.*}} invalid 'int'
   int [_, b] = {0, 0};
   [b]{};
 }
 
+namespace GH67495 {
 int get_point();
-void pr67495() {
+void f() {
   // CHECK: DecompositionDecl {{.*}} invalid 'int &'
   auto& [x, y] = get_point();
   [x, y] {};
+}
 }

